cdap-app-templates/cdap-etl/cdap-data-streams/src/main/java/co/cask/cdap/datastreams/SparkStreamingPipelineDriver.java

```diff
@@ -80,26 +82,61 @@ public void run(final JavaSparkExecutionContext sec) throws Exception {
     boolean checkpointsDisabled = pipelineSpec.isCheckpointsDisabled();
 
     String checkpointDir = null;
+    JavaSparkContext context = null;
     if (!checkpointsDisabled) {
-      // Get the location of the checkpoint directory.
       String pipelineName = sec.getApplicationSpecification().getName();
-      String relativeCheckpointDir = pipelineSpec.getCheckpointDirectory();
-
-      // there isn't any way to instantiate the fileset except in a TxRunnable, so need to use a reference.
-      final AtomicReference<Location> checkpointBaseRef = new AtomicReference<>();
-      Transactionals.execute(sec, new TxRunnable() {
-        @Override
-        public void run(DatasetContext context) throws Exception {
-          FileSet checkpointFileSet = context.getDataset(DataStreamsApp.CHECKPOINT_FILESET);
-          checkpointBaseRef.set(checkpointFileSet.getBaseLocation());
+      String baseCheckpointDir = appendToDirectory(pipelineSpec.getCheckpointDirectory(), pipelineName);
+      checkpointDir = appendToDirectory(baseCheckpointDir, pipelineSpec.getPipelineId());
+      Path checkpointDirPath = new Path(checkpointDir);
+
+      context = new JavaSparkContext();
+      Configuration configuration = context.hadoopConfiguration();
+      // Set the filesystem to whatever the checkpoint directory uses. This is necessary since spark will override
+      // the URI schema with what is set in this config. This needs to happen before StreamingCompat.getOrCreate
+      // is called, since StreamingCompat.getOrCreate will attempt to parse the checkpointDir before calling
+      // context function.
+      configuration.set("fs.defaultFS", checkpointDir);
+      // Create a URI to ensure the the directory passed in is well formed. This avoids an infinite recursion bug
+      // in the Hadoop FileSystem class.
+      try {
+        URI uri = URI.create(pipelineSpec.getCheckpointDirectory());
```

Comment:
The check is already done by the `new Path` call above. You should catch the exception up there. BTW, shouldn't the format check happens in configure time instead of execution time?

Reply:
Good point, done